go/vt/vtgate/planbuilder/builder.go

```diff
@@ -216,3 +219,101 @@ func buildVSchemaDDLPlan(stmt *sqlparser.AlterVschema, vschema ContextVSchema) (
 		AlterVschemaDDL: stmt,
 	}, nil
 }
+
+func buildFlushPlan(stmt *sqlparser.Flush, vschema ContextVSchema) (engine.Primitive, error) {
+	if len(stmt.TableNames) == 0 {
+		return buildFlushOptions(stmt, vschema)
+	}
+	return buildFlushTables(stmt, vschema)
+}
+
+func buildFlushOptions(stmt *sqlparser.Flush, vschema ContextVSchema) (engine.Primitive, error) {
+	destination, keyspace, _, err := vschema.TargetDestination("")
+	if err != nil {
+		return nil, err
+	}
+	return &engine.Send{
+		Keyspace:          keyspace,
+		TargetDestination: destination,
+		Query:             sqlparser.String(stmt),
+		IsDML:             false,
+		SingleShardOnly:   false,
+	}, nil
+}
+
+func buildFlushTables(stmt *sqlparser.Flush, vschema ContextVSchema) (engine.Primitive, error) {
+	type sendDest struct {
+		ks   *vindexes.Keyspace
+		dest key.Destination
+	}
+
+	flushStatements := map[sendDest]*sqlparser.Flush{}
+	for i, tab := range stmt.TableNames {
+		var destinationTab key.Destination
+		var keyspaceTab *vindexes.Keyspace
+		var table *vindexes.Table
+		var err error
+		table, _, _, _, destinationTab, err = vschema.FindTableOrVindex(tab)
+
+		if err != nil {
+			return nil, err
+		}
+		if table == nil {
+			return nil, vindexes.NotFoundError{TableName: tab.Name.String()}
+		}
+		keyspaceTab = table.Keyspace
+		stmt.TableNames[i] = sqlparser.TableName{
+			Name: table.Name,
+		}
+		if destinationTab == nil {
+			destinationTab = key.DestinationAllShards{}
+		}
+
+		flush, isAvail := flushStatements[sendDest{keyspaceTab, destinationTab}]
+		if isAvail {
+			flush.TableNames = append(flush.TableNames, stmt.TableNames[i])
+		} else {
+			flush = &sqlparser.Flush{
+				IsLocal:      stmt.IsLocal,
+				FlushOptions: nil,
+				TableNames:   sqlparser.TableNames{stmt.TableNames[i]},
+				WithLock:     stmt.WithLock,
+				ForExport:    stmt.ForExport,
+			}
+		}
+		flushStatements[sendDest{keyspaceTab, destinationTab}] = flush
+	}
+
+	keys := make([]sendDest, len(flushStatements))
+
+	// Collect keys of the map
+	i := 0
+	for k := range flushStatements {
+		keys[i] = k
+		i++
+	}
+
+	sort.Slice(keys, func(i, j int) bool {
+		return keys[i].ks.Name < keys[j].ks.Name
+	})
+
+	var finalPlan engine.Primitive
+	for _, sendDest := range keys {
+		plan := &engine.Send{
+			Keyspace:          sendDest.ks,
+			TargetDestination: sendDest.dest,
+			Query:             sqlparser.String(flushStatements[sendDest]),
+			IsDML:             false,
+			SingleShardOnly:   false,
+		}
+		if finalPlan == nil {
+			finalPlan = plan
+		} else {
+			finalPlan = &engine.Concatenate{
+				Sources: []engine.Primitive{finalPlan, plan},
+			}
+		}
+	}
```

Comment:
this change means it will look like if there are 3 send plans
concat2 {concat1{send1, send2}, send3}

whereas it can be planned as
concat {send1, send2, send3}

add a test containing 3 send plans.

Reply:
Concat only takes two plans. It is an array but it is implicitly assumed that it only has 2 elements

Reply:
Changed that behaviour now